README
$Id: README,v 1.5 2004/04/20 18:05:25 jeff Exp $

extproc_perl enables Oracle stored procedures and functions to be written
in Perl.  It uses Oracle's external procedure interface to create Perl
interpreters on the fly to execute Perl subroutines.  These interpreters
are persistent throughout the life of the database session.

Please send all bug reports, questions and suggestions to jeff@smashing.org.

More information can be found at http://www.smashing.org/extproc_perl

REQUIREMENTS
------------
o A version of Solaris or Linux supported by Oracle
o Perl 5.6.1 or later
o shared libperl, threading disabled (threading perl crashes extproc_perl)
o Oracle 8i (8.1.7 or later) or 9i server with OCI development libraries and
  headers (problems have been reported with versions earlier than 8.1.7)
o DBI and DBD::Oracle 1.15

DOCUMENTATION
-------------
There is a comprehensive user's guide available on the web at
http://www.smashing.org/extproc_perl/userguide.html.  This is the official
documentation for extproc_perl.

TESTING
-------
WARNING: The test suite is very immature.  Failures DO NOT indicate that
extproc_perl will not run on your system, but you should feel free to report
them.

Before you run 'make test', you need to do several things to set up your
environment, some of which you may be familiar with from DBD::Oracle:

* Set the ORACLE_SID environment variable to the name of your database.
* Set the ORACLE_USERID environment variable with your username and password
  separated by a slash (e.g. user1/foobar).
* If you are using Oracle 9.2.0 or higher, add the following line to the
  SID_DESC section of your external procedure listener configuration:

  (ENVS="EXTPROC_DLLS=/path/to/your/build/directory/extproc_perl_test.so")

  Replace /path/to/your/build/directory with the absolute path to the directory
  containing the extproc_perl source code.  You configuration should look
  something like this:

    (SID_DESC =
      (SID_NAME = PLSExtProc)
      (ORACLE_HOME = /u01/app/oracle/product/9.2.0)
      (PROGRAM = extproc32)
      (ENVS="EXTPROC_DLLS=/home/jeff/build/extproc_perl2/extproc_perl_test.so")
    )

  Reload the listener with 'lsnrctl reload' and you should be ready to test.

  IMPORTANT!!!  Remove this configuration once you're finished testing.  The
  test version of extproc_perl enables some testing features that are not
  suitable for production use.


Changes
$Id: Changes,v 1.28 2006/04/07 19:56:13 jeff Exp $

2.50	Fri Apr  7 15:51:15 EDT 2006
	* new spec parser using Parse::RecDescent
	* relinking is no longer needed for new procedures
	* "create_wrapper" removed in favor of "create_extproc"
	* cleaned up source code formatting
	* new testing infrastructure with better coverage
	* changed some default configuration options in Makefile.PL

2.01	Thu Sep 16 16:16:18 EDT 2004
	* New "reparse_subs" option enables reparsing of changed subroutines
	* Removed distracting debugging output from ora_exception
	* New upgrade script
	* The filename argument for import_perl is now optional.  It will
	  default to the name of the function with a ".pl" extension.

2.00	Tue Apr 20 13:54:35 EDT 2004
	* Added "new" method for creating new OCIDate objects
	* Added "setdate_sysdate" method for setting OCIDate objects to current
	  system date/time
	* Added "ddl_format" procedure for overriding default ddl_format
	* Updated bundled documentation

1.99_09	Thu Apr 15 13:26:05 EDT 2004
	* better runtime error reporting
	* fixed problem importing source code > 4000 bytes
	* be more forgiving with whitespace in prototypes

1.99_08 Mon Apr 12 11:57:09 EDT 2004
	* per-database configuration directives (dbname:config)
	* increase theoretical max_code_size to 4 GB
	* increase default max_code_size to 32768
	* new option to output DDL in package body format
	* added ddl_format column in perl_config view
	* fixed prototype parsing to require FUNCTION/PROCEDURE declaration and
          to allow an empty argument list
	* prototypes can now be specified in the perl source
	* removed obsolete columns from code table
	* autoload some module subroutines to reduce parsing overhead

1.99_07 Wed Feb 25 18:24:39 EST 2004
	* fixed crash in perl_config and perl_status views
	* implemented preliminary test suite for "make test"
	* fixed "always NULL" VARCHAR2 return values in directly executed code
	* fixed missing null indicators for numeric IN OUT & OUT parameters
	* removed pending support for opcode security
	* removed DBD-Oracle 1.14 patch from distribution -- 1.15 includes the
	  patched code

1.99_06 Sun Feb  1 16:55:17 EST 2004
	* ExtProc module is now built as a shared object
	* libperxsi.so no longer retains old version of ExtProc when upgrading
	* can now run 'perl -cw' to check syntax if you preload extproc_perl.so
	* create trusted code directory during installation if it doesn't exist
	* rewrote gen_module_libs so it works more reliably
	* installation will warn you when shared objects are not found
	* bootstrap failure errors are now more consistent and verbose
	* is_null and set_null methods manipulate NULL status of oracle
	  datatype objects

1.99_05 Sat Jan 17 16:59:49 EST 2004
	* support for DATE types
	* methods for manipulating DATE values in Perl
	* existing libperlxsi is no longer overwritten during installation
	* set NULL indicators for IN OUT & OUT types
	* fix inc_path_to_mflag crash on solaris

1.99_04 Sat Dec 27 18:48:07 EST 2003
	* implement NULL, LENGTH, and MAXLEN indicators for variable length
          data types like VARCHAR2
	* allow user-defined attributes in dbi_connect
	* move put_line into ExtProc and remove from ExtProc::Util
	* implement filehandle interface to put_line
	* make ExtProc object-oriented (needed for put_line filehandles)
	* correctly pass NULL parameters as undefined scalars for directly
	  executed subroutines
	* autogenerate external procedure DDL in Perl.create_wrapper
	* wrote the user's guide (currently only available online)

1.99_03 Sat Dec  6 22:51:11 EST 2003
	* disallow eval in taint mode
	* register .fini function (ELF) to call pl_shutdown at library unload
	  VERY EXPERIMENTAL, but it seems to work well on both linux & solaris
	* can register user-defined destructors using register_destructor
	* implement IN OUT and OUT parameters
	* changed Perl.create_wrapper syntax to better reflect Oracle's syntax
	* prefix C wrapper function names with EP_ to prevent name clashes
	* full documentation is now available from the user's guide

1.99_02 Tue Nov 18 11:00:11 EST 2003
	* can directly call Perl subroutines, bypassing the generic
	  func & proc wrappers
	* Perl.create_wrapper creates C wrappers for directly calling subs
	* beginnings of support for non-VARCHAR2 IN & RETURN values (just
	  integers for now)
	* perlxsi.mk and other relinking files are now installed in the
	  trusted code directory

1.99_01	Sat Nov  8 12:19:19 EST 2003
	* initial alpha release (pre-2.0)
	CHANGES BELOW ARE RELATIVE TO the 1.03 codebase:
	* globals incorporated into single global EP_CONTEXT struct to minimize
	  use of globals and lay the groundwork for future threading support
	* gen_module_libs rewritten in perl for portability & accuracy
	* all oracle functions/procedures now in "Perl" PL/SQL package
	* perl function is now Perl.func
	* perl_p procedure is now Perl.proc
	* former "special subroutines" are their own PL/SQL procedures in the
	  Perl package
	* func & proc now take up to 8 arguments by default
	* debugging and tainting support are always built in
	* new configuration file with various runtime options
	* tainting is now a runtime setting
	* bootstrap file path is a runtime setting
	* can specify additions to @INC as a runtime setting
	* can specify debug directory as a runtime setting
	* simple_query.c and simple_query functions decommissioned
	* oci.c now contains all significant OCI routines
	* ExtProc "convenience" routines removed
	* new fetch_code routine dedicated to fetching code from database
	* code in database now stored as VARCHAR2 instead of CLOB
	* support for per-session namespaces (packages)
	* renamed ExtProc::exception to ExtProc::ora_exception
	* ExtProc::is_function & ExtProc::is_procedure will tell you the
	  context of the current call
	* split module shared objects into their own library, libperlxsi,
	  so they can be maintained separately from extproc_perl.so

ExtProc/lib/ExtProc/Code.pm
# $Id: Code.pm,v 1.50 2006/04/07 19:41:09 jeff Exp $

package ExtProc::Code;

use 5.6.1;
use strict;
use warnings;
use AutoLoader 'AUTOLOAD';

require Exporter;

our @ISA = qw(Exporter);
our %EXPORT_TAGS = ( 'all' => [ qw(
    &create_extproc
    &import_code
    &drop_code
) ] );
our @EXPORT_OK = ( @{ $EXPORT_TAGS{'all'} } );
our @EXPORT = qw(
);
our $VERSION = '2.50';

use ExtProc qw(ep_debug put_line);
use File::Spec;
use DBI;
use Parse::RecDescent;

# for external procedure parameters
our %typemap = (
    'PLS_INTEGER' => 'int',
    'REAL' => 'float',
    'VARCHAR2' => 'string',
    'DATE' => 'OCIDate'
);

# type signatures
our %signature = (
    'VARCHAR2'    => 'c',
    'PLS_INTEGER' => 'i',
    'REAL'        => 'r',
    'DATE'        => 'd',
);

# a persistent Parse::RecDescent parser
our $parser;

1;

__END__

### AUTOLOADED SUBROUTINES ###

# create_extproc(spec,[lib])
# create external procedure based on spec and optional library
sub create_extproc
{
    my ($spec, $lib) = @_;

    # parse the spec
    my $attrs = parse_spec($spec);
    die "invalid specification: $spec" unless $attrs;
    my $sig = args2sig($attrs->{'returntype'}, $attrs->{'args'});

    # set function to call
    my $retsig = substr($sig, 0, 1);
    my $func = "ep_call_";
    if ($retsig eq 'v') {
        $func .= 'proc';
    }
    else {
        $func .= "func_$retsig";
    }

    # generate DDL
    my $sql;
    my $ddl_format = ExtProc::config('ddl_format');
    if ($ddl_format == 0) {
        $sql = 'CREATE OR REPLACE ';
    }
    else {
        $sql = "-- for package specification\n$spec\n";
        $sql .= "-- for package body\n";
    }
     
    $sql .= $attrs->{'calltype'};
    $sql .= " $attrs->{'name'}(";
    foreach my $arg (@{$attrs->{'args'}}) {
        my ($name, $dir, $type) = @{$arg};
        $sql .= "$name $dir $type, ";
    }
    $sql .= "ep_sub IN VARCHAR2 := '$attrs->{'name'}', ep_sig IN VARCHAR2 := '$sig')\n";

    if ($attrs->{'calltype'} eq 'FUNCTION') {
        $sql .= "RETURN $attrs->{'returntype'}\n";
    }

    $sql .= "AS EXTERNAL NAME \"$func\"\n";
    $sql .= "LIBRARY \"" . ($lib ? $lib : "PERL_LIB") . "\"\n";
    $sql .= "WITH CONTEXT\n";
    $sql .= "PARAMETERS (\n";
    $sql .= "   CONTEXT";
    if ($attrs->{'calltype'} eq 'FUNCTION') {
        $sql .= ",\n   RETURN INDICATOR BY REFERENCE"
    }

    $sql .= ",\n   ep_sub string";
    $sql .= ",\n   ep_sig string";

    foreach my $arg (@{$attrs->{'args'}}) {
        my ($name, $dir, $type) = @{$arg};
        $sql .= ",\n   $name $typemap{uc($type)}";
        $sql .= ",\n   $name INDICATOR short";
        if (uc($type) eq 'VARCHAR2') {
            $sql .= ",\n   $name LENGTH sb4";
            if ($dir =~ /OUT/i) {
                $sql .= ",\n   $name MAXLEN sb4";
            }
        }
    }

    $sql .= "\n);";
    if ($ddl_format == 0) {
        $sql .= "\n/\n";
    }

    # write DDL to file
    local *DDL;
    my $trusted_dir = ExtProc::config('trusted_code_directory');
    open(DDL, '>', File::Spec->catfile($trusted_dir, $attrs->{'name'} . '.sql' ))
        or die $!;
    print DDL "$sql\n";
    close(DDL);

    # output DDL ("set serveroutput on" to see it in sqlplus)
    #put_line($_) foreach (split(/[\r\n]+/, $sql));
}

# import_code(name, [filename], [spec])
# import code from a file in the trusted code directory and optionally create
# a C wrapper based on the supplied spec
sub import_code
{
    my ($name, $file, $spec) = @_;

    # DML -- MUST BE CALLED AS A PROCEDURE
    if (!ExtProc::is_procedure) {
        ExtProc::ora_exception('import_code must be called as a procedure!');
        return;
    }

    if ($name eq '') {
        ExtProc::ora_exception('import_code: empty subroutine name');
        return;
    }

    # untaint arguments, since we're being called from oracle
    if ($name =~ /^([A-z\d\-_]+)$/) {
        $name = $1;
    }
    else {
        ExtProc::ora_exception('illegal characters in subroutine name');
        return;
    }
    if ($file =~ /^([\w\.\-\_]+)$/) {
        $file = $1;
    }
    elsif (!defined($file)) {
        $file = "${name}.pl";
    }
    else {
        ExtProc::ora_exception('illegal characters in filename');
        return;
    }

    # what's our code table and trusted code directory?
    my $table = ExtProc::config('code_table');
    my $dir = ExtProc::config('trusted_code_directory');

    my $path = File::Spec->catfile($dir, $file);
    my $size = (stat($path))[7];
    if ($size > ExtProc::config('max_code_size')) {
        ExtProc::ora_exception("file too large for import ($size bytes)");
        return;
    }

    # read code from file
    my ($code, $line);
    local *CODE;
    open(CODE, $path) or die "failed to open code file: $!";
    while(defined($line = <CODE>)) {
        $code .= $line;
    }
    close(CODE);

    # get current version number, if any
    my $dbh = ExtProc->dbi_connect;
    my $sth = $dbh->prepare("select nvl(version, 0) from $table where name = ?");
    $sth->execute($name);
    ExtProc::ep_debug("rows=".$sth->rows);
    my $version = ($sth->fetchrow_array)[0];
    $sth->finish;

    # delete existing code if it exists
    $sth = $dbh->prepare("delete from $table where name = ?");
    $sth->execute($name);
    $sth->finish;

    # find spec, if any
    if (!$spec && $code =~ /^((?:FUNCTION|PROCEDURE)\s+$name.*)/im) {
        $spec = $1;
    }

    # import code into database, incrementing version
    $sth = $dbh->prepare("insert into $table (name, plsql_spec, language, version, code) values(?, ?, 'Perl5', ?, ?)");
    $sth->execute($name, $spec, $version+1, $code);
    $sth->finish;

    # create C wrapper if we have a spec
    $spec && create_extproc($spec);
}

# drop_code(name)
# silently remove code from code table
sub drop_code
{
    my $name = shift;

    # DML -- MUST BE CALLED AS A PROCEDURE
    if (!ExtProc::is_procedure) {
        ExtProc::ora_exception('drop_code must be called as a procedure!');
        return;
    }

    # what's our code table?
    my $table = ExtProc::config('code_table');

    my $dbh = ExtProc->dbi_connect;
    my $sth = $dbh->prepare("delete from $table where name = ?");
    $sth->execute($name);
    $sth->finish;
}

sub parse_spec
{
    my $spec = shift;

    # clear args, which may carry over from previous invocation
    @ExtProc::Parser::args = ();
    
    unless ($parser) {
        $parser = new Parse::RecDescent(q{
            datatype: 'VARCHAR2' | 'PLS_INTEGER' | 'REAL' | 'DATE'

            direction: 'IN' | 'OUT'

            name: /[\w_]+/

            argument: name direction(s) datatype
                { $return = [$item[1], join(' ', @{$item[2]}), $item[3]]; }

            arglist: '(' argument(s /,/) ')'
                { @ExtProc::Parser::args = @{$item[2]}; }

            spec2: name arglist(?)
                { $ExtProc::Parser::callname = $item[1]; }

            procedure: 'PROCEDURE' spec2
                { $ExtProc::Parser::calltype = 'PROCEDURE';
                  $ExtProc::Parser::returntype = undef;
                  1; }

            function: 'FUNCTION' spec2 'RETURN' datatype
                { $ExtProc::Parser::calltype = 'FUNCTION';
                  $ExtProc::Parser::returntype = $item[4]; }

            spec: procedure | function

            startrule: spec
        });
    }

    my $res = $parser->startrule($spec);
    return undef unless $res;

    return ({
        calltype   => $ExtProc::Parser::calltype,
        returntype => $ExtProc::Parser::returntype,
        name       => $ExtProc::Parser::callname,
        args       => [@ExtProc::Parser::args]
    });
}

sub args2sig
{
    my ($returntype, $argref) = @_;
    my @args = @{$argref};
    my $sig = (defined $returntype) ? $signature{$returntype} : 'v';
    foreach my $i (0..$#args) {
        if (uc($args[$i]->[1]) eq 'OUT') {
            $sig .= "O$signature{$args[$i]->[2]}";
        }
        elsif (uc($args[$i]->[1]) eq 'IN OUT') {
            $sig .= "B$signature{$args[$i]->[2]}";
        }
        else {
            $sig .= "I$signature{$args[$i]->[2]}";
        }
    }
    return $sig;
}
